CcdhModel<-->TerminologyService integration via GH Actions

joeflack4 opened the pull request:
## Updates
1. Added a file to: (i) trigger rebuild of terminology service, which then causes (ii) a rebuild of the CCDH model.
2. Added a placeholder file to: Just rebuild the CCDH model only.  _(@Gaurav: I left this largely empty. I believe you mentioned you will fill it out after.)_

## Thoughts
There's a lot of flexibility with how we want to implement this integration. I don't really know the intricacies of how these two repos depend on each other. But in terms of order of refreshing, what I've added supports the following multi-repository workflows:

### Workflow 1
1. CcdhModel: `refreshTermServiceAndCcdhModel.yml` is run manually. It doesn't build model repo. It just calls terminology service file.
2. TerminologyService: `rebuildProductionServer.yml` gets called
3. TerminologyService: As the build script is running, `importer.py` gets called. At the end of that script, it sends a `repository_dispatch` event to the CcdhModel repository.
4. CcdhModel: `refreshModel.yml` is set up as the only file with a `repository_dispatch` trigger, so it is automatically the file that gets run when the `repository_dispatch` from the TerminologyService is made.

### Workflow 2
This has the same effects as "Workflow 1". The only difference is how the workflow is triggered.
1. TerminologyService: `rebuildProductionServer.yml` is run manually
2. TerminologyService: As the build script is running, `importer.py` gets called. At the end of that script, it sends a `repository_dispatch` event to the CcdhModel repository.
3. CcdhModel: `refreshModel.yml` is set up as the only file with a `repository_dispatch` trigger, so it is automatically the file that gets run when the `repository_dispatch` from the TerminologyService is made.

---

I could imagine other ways we could do this, though. I'm not sure if how I've set this up is the best way / the way you guys would like to do this.

I'm wondering if, for example, we would like **step 1** to be a build of the CcdhModel before the terminology service is built. 

Another thing we could do differently, if we needed to, is that (i) the terminology service is built, (ii) it triggers the model to build itself, and then (iii) the model triggers the terminology service to update again. Is that something we need to do? This can be done. However, we just need to make sure that these two repositories don't keep triggering refreshes of each other over and over again indefinitely. 

### A word about `event_type`
The `repository_dispatch` events have a param called `event_type`. In a GH action yml file, it looks like this:

```
on:
  workflow_dispatch:
  repository_dispatch:
    types: [my_event_a, my_event_b]
```

Based on my reading, `event_type`s are the only way to be able to choose which GH action yml files are run when a `repository_dispatch` is called. If no `event_type` is specified, or if the yml files don't list a `types: ` field, all of the yml files with a `repository_dispatch` are run. But adding `event_type`s gives the flexibility for us to specify 1 or more yml files to run.

I have not added any `event_type`s as of yet. But I imagine this could come in handy for us.

## Additional steps needed: adding a secret **(important)**
We need to add a secret to either the CcdhModel repository, or the cancerDHC organization. It needs to be a personal access token. It can be any of our accounts that has write access to both repositories. Perhaps we can use a token from our new user "ccdh.terminology.docker@gmail.com". This account has been set up to have access to the TerminologyService repository. I can't remember yet if Dazhi has given it write permissions for the ccdhmodel repo or at the cancerDHC organization level.

The secret that needs to be added should be called: `GH_LIMITED_PERSONAL_ACCESS_TOKEN`. At least, that's what I've called it in `refreshTermServiceAndCcdhModel.yml`. From what I've heard, the token that needs to be created (see generally: [creating a token](https://docs.github.com/en/github/authenticating-to-github/keeping-your-account-and-data-secure/creating-a-personal-access-token)) only needs to have the `repo: status` checkbox enabled.

## Final thoughts
After merging this PR and setting up this access token, the integration should be fully set up, though we'll need to test it. However that may also prove difficult, as the terminology services build is currently failing. @jiaola I'm still working on that. I found another issue which I submitted a fix for today, but there are still some issues in one of the dependencies (`SSSOM.py`).

joeflack4 commented:
Dazhi and I spoke, and it looks like we're probably be going to set up a test server for the terminology service soon. Given that, this action will need to do a run on the test server first. The easiest way to do that will probably involve editing this GH action workflow further.

gaurav commented:
This looks good to me! I think Workflow 1 is probably going to be what we want. Here's how I imagine this working:
1. EITHER:
   1. Somebody modifies the model in Google Sheets and then runs a GitHub action (TODO) to regenerate the model from the Google Sheet, which automatically creates a pull request (PR), OR
   2. Somebody modifies the model directly at `./src/schema/ccdhmodel.yaml` and creates a PR.
2. Creating the PR automatically triggers `refreshTermServiceAndCcdhModel.yml`. Note that at this point, no permissible values have been populated into the model, but it includes all the mappings, which is what the terminology service needs.
3. We then go into "Workflow 1" as described above: `refreshTermServiceAndCcdhModel.yml` calls the TerminologyService's `rebuildProductionServer.yml`, which calls CcdhModel: `refreshModel.yml` via a repository_dispatch trigger once it is done.
4. `refreshModel.yml` will fill in the permissible values for all enumerations in the model via the Terminology Service's API and update the PR to (1) include the updated model, and (2) to regenerate all the artifacts.
5. Any further changes to the PR will trigger `refreshTermServiceAndCcdhModel.yml` again, and we go back to step 2.

> I'm wondering if, for example, we would like step 1 to be a build of the CcdhModel before the terminology service is built.

This shouldn't be necessary, since in my workflow above `refreshTermServiceAndCcdhModel.yml` doesn't get triggered until after the model schema at `./src/schema/ccdhmodel.yaml` is ready for use.

> I have not added any event_types as of yet. But I imagine this could come in handy for us.

I think we should absolutely do this -- even if we never add another event type, using an event type will still make it clear exactly what is being triggered here. Good find, @joeflack4!

> Perhaps we can use a token from our new user "ccdh.terminology.docker@gmail.com". This account has been set up to have access to the TerminologyService repository. I can't remember yet if Dazhi has given it write permissions for the ccdhmodel repo or at the cancerDHC organization level.

This sounds good to me! I think it would be easiest to have a single bot user for all the repos. @jiaola @balhoff @turbomam What do you think?

> After merging this PR and setting up this access token, the integration should be fully set up, though we'll need to test it. 

Since it's pretty easy for me to do this manually when DMH needs a new version of the model deployed, I don't think this is very urgent -- I plan to review this PR in detail around the start of August, and then merge it after review once it's good to go. We can figure out the bot and access token by then as well. What do you think, @joeflack4?

> The easiest way to do that will probably involve editing this GH action workflow further.

In that case, it might be a good idea to [mark this PR as a draft](https://github.blog/changelog/2020-04-08-convert-pull-request-to-draft/) until all the terminology service side of this is up and running.

joeflack4 commented:
Hey @gaurav, thanks for the detailed response. Overall, this sounds like a good plan to me!

## Multi-action workflows
Regarding (workflow pt 2), _triggering `refreshTermServiceAndCcdhModel.yml` on PR_, this sounds good and it is a simple change for me to make, but I have a question and concern. I don't think this is a concern if the YML file is updated manually, but in the case of Google sheets, I suppose we will need to ensure that these get up actions do not trigger at the same time (as they would in workflow 1 below)?

I'm assuming for now that  the `refreshTermServiceAndCcdhModel.yml`, `refreshModel.yml` and the action that builds from google sheets, e.g. `buildFromGoogleSheets.yml` should be 3 separate actions. But I just want to throw out the possibility that, if we wanted, I imagine we could combine 2-3 of these files into a single action if we wanted. We would just need to specify certain params and pass those in when calling it in order to control which steps it executes and make sure it didn't keep calling itself in an infinite loop. I'm imagining that separate is better, though.

In any case, based on what you are saying, here are my new imagined multi-action workflows (differences in **bold**). I prefer "workflow 2".

### Workflow 1: PR triggers
**(i) We have a GH action `buildFromGoogleSheets.yml` on PR only, and `refreshTermServiceAndCcdhModel.yml` is on PR only. They're both triggered at the same time.
(i.2) Hopefully `buildFromGoogleSheets.yml` finishes before `refreshTermServiceAndCcdhModel.yml`-->`rebuildProductionServer.yml` gets far enough to utilize the update.**
(iii) Terminology service runs modified `rebuildProductionServer.yml` to build on test server, then on production, and then calls an action to build the ccdh model again,  `refreshModel.yml`.
(iv)`refreshModel.yml` runs

### Workflow 2: PR and `repository_dispatch` triggers
**(i) We have a GH action `buildFromGoogleSheets.yml` on PR or on manual workflow_dispatch, and the last step of that action uses `repository_dispatch` event to call `refreshTermServiceAndCcdhModel.yml`-->``rebuildProductionServer.yml`.**
(iii) Terminology service runs modified `rebuildProductionServer.yml` to build on test server, then on production, and then calls an action to build the ccdh model again,  `refreshModel.yml`.
(iii)`refreshModel.yml` runs

### Workflow 3: PR trigger & calling action from another action
**(i) We have a GH action `buildFromGoogleSheets.yml` on PR or on manual workflow_dispatch, and the last step of that action uses calls `refreshTermServiceAndCcdhModel.yml`. The only difference between this and "workflow 2" is that we would use a custom action like https://github.com/convictional/trigger-workflow-and-wait to call the file instead of using a `repository_dispatch`. There are several such custom actions to trigger actions. I just want to note that this particular one I linked has the ability to "wait" for the called action to complete, and can return control back to the calling action. I was originally going to use this instead of `repository_dispatch`, and this could be something interesting to explore if it is preferred, but so far I have had difficulty with getting `rebuildProductionServer.yml` to know when everything on the server has finished.**
(iii) Terminology service runs modified `rebuildProductionServer.yml` to build on test server, then on production, and then calls an action to build the ccdh model again,  `refreshModel.yml`.
(iii)`refreshModel.yml` runs

## `workflow_dispatch` triggers
I suppose I'll keep the `workflow_dispatch` trigger enabled for `refreshTermServiceAndCcdhModel.yml`  just in case we want to run it manually but feel free to remove it if that's better.

## `event_type` and `repository_dispatch` triggers
My intuition is with yours here. I'll go ahead and have each file accept only 1 unique `event_type`. It looks like `event_type` can have spaces, but maybe it's easier for me to just name the `event_type`s after the exact file they're referring to. E.g. `refreshModel.yml`'s `event_type` can just be "refreshModel".

## Single bot user for all repos
That sounds good to me. I've set up a gmail, GitHub account, ccdh-terminology-service server user, and corresponding secrets to allow "ccdh.terminology.docker@gmail.com" to function for our purposes. But this can easily be re-done for a different bot. You guys just discuss and I'll pitch in if/where needed.

If we have a single bot for all repos, then it may be better for us to put any shared secrets that the bot would use at the organization level, rather than repeating said secrets in multiple repositories.

## Start of August, PR review, & final steps to set up any additional repository secrets
Sure thing, I can wait!

joeflack4 commented:
@gaurav Hey Gaurav, just checking ETA on this.

gaurav reviewed (approved):
Looks good to me! Given that I still need to make changes here, I think the right move is to merge this code in its current state, and then I'll update it in another PR to get refreshModel and buildFromGoogleSheets set up and working correctly, and then we can try out the whole workflow and see if it works.

(Incidentally, @joeflack4, I think you're a developer on this repository, so feel free to make new branches directly in this repository rather than creating your own branch if you'd prefer! I'd just ask that you submit a PR for review by Jim Balhoff or me rather than merging the change directly into the `main` branch.)

gaurav commented:
@joeflack4 Oof, thanks for the reminder -- this totally fell off my radar last month!

> I'm assuming for now that the `refreshTermServiceAndCcdhModel.yml`, `refreshModel.yml` and the action that builds from google sheets, e.g. `buildFromGoogleSheets.yml` should be 3 separate actions. But I just want to throw out the possibility that, if we wanted, I imagine we could combine 2-3 of these files into a single action if we wanted. We would just need to specify certain params and pass those in when calling it in order to control which steps it executes and make sure it didn't keep calling itself in an infinite loop. I'm imagining that separate is better, though.

I agree that separate is better, especially as we start this. It probably makes sense to group them into a single YAML file eventually. I'm also struggling to think of a case where `refreshTermServiceAndCcdhModel.yml` will need to be called except right after `buildFromGoogleSheets.yml`, so _maybe_ we could merge those two. But let's start it simple and make it more complex later :-)

Also, I think that our hope is that _eventually_ most of the model design work will be done and we can make incremental changes to the YAML schema directly, but unless we can improve LinkML editor support in the meantime, I don't think that's going to happen for a while yet.

> If we have a single bot for all repos, then it may be better for us to put any shared secrets that the bot would use at the organization level, rather than repeating said secrets in multiple repositories.

That makes sense. Could you please write up the information on the bot (what tasks it carries out, what triggers it, etc.) and document it in one of our repositories? I'd like to make sure that we document why we created this bot and what it's intended for. It might also be a good idea to e-mail the bot credentials to Moni, just so we have a remote backup of those credentials just in case a meteor hits Baltimore or something :-P

> I prefer "workflow 2".

Workflow 2 looks good to me too! The key piece of work that I'll need to do is separate the Google Sheet generator into a Google Sheet generator component (that will be called from `buildFromGoogleSheets.yml`), which will generate the CRDCH model from the Google Sheet, and a Refresh Model component (that will be called from `refreshModel.yml`), which will only regenerate the valuesets in the model by calling the Terminology Service.

So, next steps: I notice that you haven't created a `buildFromGoogleSheets.yml` -- if you'd like to do that before merging this PR, go ahead. Otherwise, I've approved this PR -- you can go ahead and merge it now, and I'll fill out the remaining YAML files in a separate PR. Whatever works best for you!

(One question: why `.yml` instead of `.yaml`?)

joeflack4 commented:
NP @gaurav . Thanks for confirming the workflow with me. It looks like we've come to an agreement as to what will work for now. I went ahead and invited the current docker bot to the ccdh model repository and I also added the necessary secret.

I think that's a good idea about creating an issue for the organization controlled bot and organization level secrets. I made an issue for it under 'operations': https://github.com/cancerDHC/operations/issues/96.

Regarding `buildFromGoogleSheets.yml`, I added it as a blank fire. I'll leave it to you to update this and the `refreshModel.yml`.

Concerning yaml vs yml, I don't really have a preference. Dazhi just had the our repo already set up with a `main.yml` and I just followed that convention.

Regarding forks, this is just how I've done things at previous organizations. There are some minor benefits, such as being able to have my own set of git tags. But I can also change this and make branches directly on the trunk repository if people prefer.

Thanks for approving this PR. Now that I've made the small changes and created the GitHub issue, I'll go ahead and merge it!
